src/consumer/base_consumer.rs

```diff
@@ -96,66 +63,181 @@ where
-            }
-            timeout -= op_timeout;
-        }
-    }
-
     /// Polls the consumer for new messages.
     ///
     /// It won't block for more than the specified timeout. Use zero `Duration` for non-blocking
     /// call. With no timeout it blocks until an event is received.
     ///
     /// This method should be called at regular intervals, even if no message is expected,
-    /// to serve any queued callbacks waiting to be called. This is especially important for
+    /// to serve any queued events waiting to be handled. This is especially important for
     /// automatic consumer rebalance, as the rebalance function will be executed by the thread
     /// calling the poll() function.
     ///
     /// # Lifetime
     ///
     /// The returned message lives in the memory of the consumer and cannot outlive it.
     pub fn poll<T: Into<Timeout>>(&self, timeout: T) -> Option<KafkaResult<BorrowedMessage<'_>>> {
-        self.poll_raw(timeout.into())
-            .map(|ptr| unsafe { BorrowedMessage::from_consumer(ptr, self) })
+        self.poll_queue(self.get_queue(), timeout)
+    }
+
+    pub(crate) fn poll_queue<T: Into<Timeout>>(
+        &self,
+        queue: &NativeQueue,
+        timeout: T,
+    ) -> Option<KafkaResult<BorrowedMessage<'_>>> {
+        let mut timeout = timeout.into();
+        let min_poll_interval = self.context().main_queue_min_poll_interval();
+        loop {
+            let op_timeout = std::cmp::min(timeout, min_poll_interval);
+            let maybe_event = self.client().poll_event(queue, op_timeout);
+            if let Some(event) = maybe_event {
+                let evtype = unsafe { rdsys::rd_kafka_event_type(event.ptr()) };
+                match evtype {
+                    rdsys::RD_KAFKA_EVENT_FETCH => {
+                        if let Some(result) = self.handle_fetch_event(event) {
+                            return Some(result);
+                        }
+                    }
+                    rdsys::RD_KAFKA_EVENT_ERROR => {
+                        if let Some(err) = self.handle_error_event(event) {
+                            return Some(Err(err));
+                        }
+                    }
+                    rdsys::RD_KAFKA_EVENT_REBALANCE => {
+                        self.handle_rebalance_event(event);
+                    }
+                    rdsys::RD_KAFKA_EVENT_OFFSET_COMMIT => {
+                        self.handle_offset_commit_event(event);
+                    }
+                    _ => {
+                        let buf = unsafe {
+                            let evname = rdsys::rd_kafka_event_name(event.ptr());
+                            CStr::from_ptr(evname).to_bytes()
+                        };
+                        let evname = String::from_utf8(buf.to_vec()).unwrap();
+                        warn!("Ignored event '{}' on consumer poll", evname);
+                    }
+                }
+            }
+
+            if op_timeout >= timeout {
```

Comment:
We set `op_timeout` to `std::cmp::min(timeout, min_poll_interval)`, so it's never going to be larger than `timeout`. Shouldn't we do something like:

```rust
let now = Instant::now();
let op_timeout = std::cmp::min(timeout, min_poll_interval);
// ...
let timeout = timeout.saturating_sub(now.elapsed());
if timeout == 0 {
    return None;
}
```

Reply:
It seems to work fine for the following common cases. Did I miss something?

- If the user provides a `timeout` set to 100ms, then the `op_timeout` will be 100ms (as `min_poll_interval` is 1sec by default). If no msgs or errors have been returned, then on this check `op_timeout` is equal to `timeout` and we return `None`.
- If the user provides a `timeout` set to 2 seconds (over the default for `min_poll_interval`), then `op_timeout` will be 1 second. Let's say we poll for 1 second and it does not return a msg/error, then `op_timeout` is less than timeout and now we do `timeout -= op_timeout` which set `timeout` to 1 second. If we get None messages/errors again, then `op_timeout` is equal to `timeout` and we return `None`.
- If the user provides `timeout` set to Never, then `op_timeout` is set to the default of 1 sec and we keep polling every second until we get a message or error.

Reply:
I was concerned about something like the following:

* the user sets the `timeout` to 100ms. 
* After 1ms, we get `RD_KAFKA_EVENT_OFFSET_COMMIT` event. 
* We do `timeout -= op_timeout`, which puts `timeout` at 0. 
* We exit.

Reply:
ah that's a good catch. Ok, I'll fix this.